designs/2022-parsing-session-objects/README.md

```diff
@@ -0,0 +1,247 @@
+
+Using the session proposed in this RFC to determine whether it's in a "single" session mode _without_ `--fix` would allow typescript-eslint to fix those bugs (https://github.com/typescript-eslint/typescript-eslint/pull/6178).
+
+## Detailed Design
+
+This RFC proposes a `session` object be provided by ESLint to parsers that at first contains just two pieces of information:
+
+```ts
+type SessionMode = "persistent" | "single";
+
+interface LintSession {
+  mode?: SessionMode;
+  options: ESLintOptions;
+}
+```
+
+The expectation for this session object is that consumers may provide a `mode` _if_ they know it.
+Parsers would then receive the session in their [`parseForESLint`](https://eslint.org/docs/latest/developer-guide/working-with-custom-parsers) as a third parameter.
+
+```js
+parser.parseForESLint(textToParse, parserOptions, lintSession);
+```
+
+### Code Flow
+
+Consumers of ESLint today use up to four classes from ESLint:
+
+- `Linter`: A lower-level class used when `fs` is not available, such as in browsers.
+- `CLIEngine` _(legacy)_: [no longer exported as of ESLint v8.0.0](https://eslint.org/docs/latest/user-guide/migrating-to-8.0.0#remove-cliengine); API predecessor to `ESLint`.
+- `ESLint`: The primary class to use in Node.js applications that includes `fs` operations.
+- `FlatESLint`: A preview of `ESLint` that uses the new [flat config](https://github.com/eslint/rfcs/pull/9)
+
+For example, the [ESLint extension for VS Code](https://github.com/Microsoft/vscode-eslint) [uses either `CLIEngine` or `ESLint`](https://github.com/Microsoft/vscode-eslint/blob/f095ba58a6bbefd5b7e5ddcff778f3f04551f554/server/src/eslint.ts#L1017-L1025), depending on the ESLint version and its configuration.
+
+#### `Linter`
+
+In `lib/linter.js`, the standalone `parse` function is what calls to `parseForESLint`.
+It is called within the `Linter` class.
+That `Linter` class can be given two additions to support passing a lint session object to `parse`:
+
+- A third property on its config object, `session`
+- The addition of that `session` object to the stored `internalSlotsMap` data for the linter
+
+The `Linter` class will also enforce the invariant assumed by the session object when `mode` is `"single"`.
+I.e. if you're running single mode, then each file MUST be linted exactly once (unless fixed).
+
+```ts
+const linter = new ESLint({ session: { mode: 'single' },  ... });
+
+// No errors on these lines
+linter.lintFiles('foo.ts');
+linter.lintFiles('bar.ts');
+
+// Error: Cannot lint the same file twice without fixes in 'single' mode
+linter.lintFiles('foo.ts');
+```
+
+#### `CLIEngine`
+
+The `CLIEngine` class is still used internally by the `ESLint` class to create a new `Linter`.
+That means it cannot assume it is being run by the CLI.
+It therefore needs to receive its session information as a constructor parameter.
```

Comment:
Aside: this is why I filed [Change Request: Rename internal CLIEngine to ESLintEngine or similar? #16670](https://github.com/eslint/eslint/issues/16670). A little confusing to work on this RFC 😄